robot/packages/package_path_visitor.go

```diff
@@ -1,22 +1,55 @@
 package packages
 
 import (
+	"fmt"
 	"reflect"
+	"strings"
 )
 
 // PackagePathVisitor is a visitor that replaces strings containing references to package names
 // with the path containing the package files on the robot.
 type PackagePathVisitor struct {
 	packageManager Manager
+	packagePaths   map[string]string
 }
 
 // NewPackagePathVisitor creates a new PackagePathVisitor.
-func NewPackagePathVisitor(packageManager Manager) *PackagePathVisitor {
+func NewPackagePathVisitor(packageManager Manager, packagePaths map[string]string) *PackagePathVisitor {
 	return &PackagePathVisitor{
 		packageManager: packageManager,
+		packagePaths:   packagePaths,
 	}
 }
 
+func (v *PackagePathVisitor) getExpectedFilepathForPackagePlaceholder(placeholder string) (string, error) {
+	filepath, ok := v.packagePaths[placeholder]
+	if !ok {
+		return placeholder, fmt.Errorf("there is no corresponding real path for this valid placeholder %s", placeholder)
+	}
+	return filepath, nil
+}
+
+// VisitAndReplaceString replaces a string with a package path if its a valid package placeholder.
+func (v *PackagePathVisitor) VisitAndReplaceString(s string) (string, error) {
+	if len(s) > 0 && s[0] != '$' {
+		// don't error here as its not a placeholder
+		return s, nil
+	}
```

Comment:
Can this be skipped and rely on the specific error from the PlaceholderPath call? That means it gets tested by the regex, rather than this primitive check here, and thus there's only one place to edit if the placeholder format changes. Also, it means you could have replacements midstring, which may be important for future use cases.

Reply:
So the reason that I added it in here was that because this is going through all strings in an attribute map (which can validly not be placeholders), we don't want to error saying string is not a valid placeholder because often its not trying to be a placeholder. We know that any placeholder has to have the ${} so doing this check will stop a lot of errors for normal filepaths that are correct.

Reply:
So... just don't error on those cases? If it matches a full placeholder regex, it's a placeholder, if it doesn't, it's not. Only errors would be for a placeholder that's syntactically valid but not in the packages config. Otherwise, this is effectively assuming anything starting with "$" is meant to be a placeholder, which sort of defeats the purpose of having the complex syntax with the brackets. Generally that sort of complex pattern is meant to avoid accidentally matching simple uses of special symbols $. Likewise, if we ever need a placeholder to NOT be the very front of a string (my previous example of `python3 ${packages.modules.myPyModule}`) this wouldn't work either.

In other words, I think we want something to be able to have a config attribute that is `cost: "$100.00"` and not get an error on it.

All that said, I won't block on this if you feel very strongly still.

Reply:
I see your point about the $ sign use case and the executable. So the current replacement logic does match at the beginning of a string and does this same check, so I am not too attached to having it remain or to change it later. 

That being said I am about to take vacation for a week so would like to get this in before then ... Changing it to not return an error would require some more work to the placeholder check because right now it will not differentiate an error between a badly formed placeholder and just a regular string.